Trim coin cache by coin height

ptschip opened the pull request:
According to @gandrewstone 's recent study, newer coins are more likely to be mined than older coins.  So, if the coins cache is full then we'll trim older coins first.

ptschip committed:
Introduce nBestCoinHeight and track it's value in each coin cache

sickpig commented:
link to @gandrewstone's result
https://medium.com/@g.andrew.stone/transaction-input-size-optimization-152cdad1b76f

ptschip committed:
Have SaltedOutPointHasher return a uint64_t

We no longer use a boost map therefore we can return a uin64_t
on all systems rather than size_t.  This was previously an issue
on a narrow range of boost versions which resulted in failures when syncing
the chain (Core issue #4634).

ptschip committed:
Trim coins cache by coin height

The cache is trimmed by the coin height.  If the height is not
high enough to trim any values then we re-adjust the trim height
up towards the current chain tip until we trim enough coins. If
there are still not enough coins to be trimmed then we just trim
any coins than are not dirty.

Futhermore, once the coin height has moved closer to the tip and
we end up not needing to trim any coins when trim is called, then
we gradually walk the trim height away from the tip so when trim
is once again succesful we will potentially maintain more of the
recent in cache.

This process of trimming by height only comes into effect when
the cache limit is exceeded and is therefore only used by more
marginal nodes that do not have a large dbcache.

ptschip committed:
formatting

ptschip committed:
Remove unnecessary iterator definition

ptschip committed:
Simplify the calculation of nTrimHeightDelta

stckwok commented:
I am trying to get this pull request changes to my local branch and then run unit and qa test.
I am not able to get the latest changes from the dev branch
git checkout dev
git pull
Already up-to-date

stckwok commented:
Thanks Peter, 
I am able to see the latest changes after adding dev_trim branch
git remote add ptschip https://github.com/ptschip/bitcoin.git
git fetch ptschip dev_trim
git checkout dev_trim
git log

commit 24b1f1b4562ba3610b7f3b4140beb6732bf9c1bd
Author: Peter Tschipper <peter.tschipper@gmail.com>
Date:   Fri Dec 8 06:53:13 2017 -0800

stckwok commented:
Hi Peter, 
Error is not on this pull request, but I got build failed using the dev_trim branch which you are trying to merge to dev.   
In the release branch, std::string CopyrightHolders() is not included in util.cpp. So it compiled ok.

util.cpp: In function ‘std::__cxx11::string CopyrightHolders(const string&)’:
util.cpp:848:53: error: ‘COPYRIGHT_HOLDERS’ was not declared in this scope
    std::string strCopyrightHolders = strPrefix + _(COPYRIGHT_HOLDERS);
util.cpp:851:64: error: ‘COPYRIGHT_HOLDERS_SUBSTITUTION’ was not declared in this scope
         strCopyrightHolders = strprintf(strCopyrightHolders, _(COPYRIGHT_HOLDERS_SUBSTITUTION));
                                                                ^
Makefile:6045: recipe for target 'libbitcoin_util_a-util.o' failed

Should they be defined in clientversion.h.   Just files mismatched?
Thanks

stckwok commented on src/coins.cpp:
```diff
@@ -265,22 +283,84 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
+        // Gradually increase the nTrimHeight if we didn't trim enought entries.
+        if (fDone && DynamicMemoryUsage() > nTrimSize && nTrimHeightDelta > nSmallestDelta)
+        {
+            if (nTrimHeightDelta <= nSmallestDelta * 100)
```
Compiler warning when comparing unsigned and signed integer values
uint64_t nTrimHeightDelta
int  nSmallestDelta

ptschip replied:
thx...fixed.

stckwok replied:
Peter, thanks for the fix. So I did both unit and regression tests

There are failed tests on the regression suite. They may not be related to this pull request
I also ran the regression test on "release" branch, only 1 test "excessive.py  --extended" failed.
Want to confirm if you also got similar failed tests. 
Should I create a separate ticket to track the failed test on regression suite?
Thanks

(1) I ran unit test (coins_tests) and all passed
>test_bitcoin --log_level=all --run_test=coins_tests

(2) Run the regression test suite with qa/pull-tester/rpc-tests.py
Runtime: 214 s
28 test(s) passed / 8 test(s) failed / 36 test(s) executed
4 test(s) disabled / 0 test(s) skipped due to platform

(3) Run all possible tests with qa/pull-tester/rpc-tests.py -extended
Runtime: 16012 s
50 test(s) passed / 6 test(s) failed / 56 test(s) executed
7 test(s) disabled / 0 test(s) skipped due to platform

ptschip committed:
Make nSmallestDelta a uint64_t to fix compiler warning

stckwok commented:
[rpc-tests-pullreq-812-extended.log](https://github.com/BitcoinUnlimited/BitcoinUnlimited/files/1553842/rpc-tests-pullreq-812-extended.log)
[rpc-tests-pullreq-812-noext.log](https://github.com/BitcoinUnlimited/BitcoinUnlimited/files/1553843/rpc-tests-pullreq-812-noext.log)

stckwok commented on src/coins.cpp:
```diff
@@ -265,22 +283,84 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
+                           "trim height:%d\n",
+            cachedCoinsUsage, DynamicMemoryUsage(), nTrimSize, nBestCoinHeight, nTrimHeight);
+
+        iter = cacheCoins.begin();
```
Minor :  is this call duplicate of line 313

stckwok commented:
Problem was due to my environment
1. Python 2.7 is used instead of 3
2. bitcoind is not latest. 

Did clean build, regression tests all passed (confirmed with 2 runs)

stckwok reviewed (commented):
changes good to be merged

stckwok reviewed (approved):
Execute the follow test cases, and the expected behavior is observed.

Objectives:
Marginal peers (such as win32 clients or Raspberry Pi, which are lack of memory to keep all coins in cache)
then coin cache start to get trimmed on the older coins in cache. 
Study from Andrew.S found that newer coins are more likely to be mined than older coins in cache

Preconditions:
1. Install bitcoind with changes of this pull request
2. Restrict the memory by setting a low dbcache (use 1GB) 
3. Enable debug log on "coindb" component
4. Create a data folder named "data_mainnet" to capture debug.log
5. Ensure VM has at least 150 GB of free diskspace (since need to Sync with Mainnet)

Steps:
1. Launch bitcoind with
> bitcoind -datadir=/home/sk10/data_mainnet -dbcache=1000 -debug=coindb -daemon

2. Tail debug.log to verify trimming occurs when block height reaches around 230K
> tail -f debug.log | grep rim

Note: Trim function only gets called if/when the dbcache is full

3. Verify log shows with increasing trim height values
"Re-adjusting trim height to 43278"
"Trimmed 306014 from the CoinsViewCache, current size after trim: 4875999 and usage 3936656 bytes"

Example, trim height keeps increasing 58814, 71029, 81029, 91029, 101029, 111029, .... 154069 .. 184069, 194069

4. Repeat test with lower dbcache value (500, and then 300)

5. Verify trimming start to occurs sooner than the previous block height of 230K
Execute this command to verify block height at the time trimming start to occur
> bitcoin-cli -datadir=/home/sk10/data_mainnet  getinfo

6. Verify trimming occurs more often when step 5 is reached. Timestamps between trimming log is shorten comparing to dbcache=1000

Note: lower the dbcache value, the sooner and faster the trimming will occur

gandrewstone commented on src/coins.cpp:
```diff
@@ -265,22 +283,84 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
+                nTrimHeightDelta -= (nSmallestDelta * 200);
+
+            nTrimHeight = nBestCoinHeight - nTrimHeightDelta;
+
```
Use saturating arithmetic: I'd like to see logic that ensures that nTrimHeightDelta and nTrimHeight stop at zero rather than wrap around and cause weird behavior.

ptschip replied:
ok , done...

gandrewstone commented on src/coins.cpp:
```diff
@@ -265,22 +283,84 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
+    bool fDone = false;
+    uint64_t nSmallestDelta = 50; // number of blocks to adjust trim height by
     CCoinsMap::iterator iter = cacheCoins.begin();
+    while (!fDone && DynamicMemoryUsage() > nTrimSize)
```
If someone passes a nTrimSize of 0 could we loop forever?  I'd like to see a DbgAssert which asserts in debug mode for any nTrimSize less than N, and sets it to N in release build.

What's a good N?  A couple of MB?

ptschip replied:
This can't happen.  But I added a test to coins_tests.cpp to occasionally trim a random cache to zero so that we never introduce such a problem in the future.

ptschip committed:
Use saturating arithmetic for nTrimHeightDelta and nTrimHeight

ptschip committed:
Add Trim test to coin_tests.cpp

Every once in a while trim a random cache to zero. This tests that
we never get into any kind of a spin loop.
